File: Src/Xultaik.Graphics/SwapChain.cs
// Copyright (c) Faber Leonardo. All Rights Reserved. https://github.com/FaberSanZ
// This code is licensed under the MIT license (MIT) (http://opensource.org/licenses/MIT)



using System;
using System.Collections.Generic;
using System.Text;
using Vortice.Direct3D;
using Vortice.DXGI;
using static Vortice.DXGI.DXGI;

using static Vortice.Direct3D12.D3D12;
using System.Runtime.InteropServices;
using Vortice.Direct3D12;

namespace Xultaik.Graphics
{
    public class SwapChain
    {
        public RenderDescriptor Description { get; private set; }

        //public Texture BackBuffer { get; set; }
        public Texture[] BackBuffers { get; set; }
        public Texture BackBuffer => BackBuffers[BackBufferIndex];

        public int BackBufferIndex { get; private set; } = 0;




        internal IDXGISwapChain3 NativeSwapChain;
        internal GraphicsDevice GraphicsDevice;
        private int bufferCount = 2;

        public SwapChain(GraphicsDevice graphicsDevice)
        {
            GraphicsDevice = graphicsDevice;
            Description = graphicsDevice.NativeParameters;

            CreateSwapChain();

            //BackBufferIndex = NativeSwapChain.CurrentBackBufferIndex;



            BackBuffers = new Texture[bufferCount];

            for (int i = 0; i < bufferCount; i++)
            {
                BackBuffers[i] = new Texture(GraphicsDevice);
                BackBuffers[i].InitializeFromImpl(NativeSwapChain.GetBuffer<ID3D12Resource>(i));
            }


            //BackBuffer = new Texture(GraphicsDevice);
            //BackBuffer.InitializeFromImpl(NativeSwapChain.GetBuffer<ID3D12Resource>(BackBufferIndex));
        }



        private void CreateSwapChain()
        {

            //switch (Description.Settings.Platform)
            //{
            //    case Platform.Win32:
            //         CreateSwapChainForDesktop();

            //    case Platform.UWP:
            //    default:
            //        throw new ArgumentException();

            //}


            CreateSwapChainForDesktop();

        }

        public void Present()
        {

            if (!Description.Settings.AllowTearing && !Description.Settings.VSync)
            {
                NativeSwapChain.Present(0, PresentFlags.None); // TODO: Defaul Full Screen ?
            }


            if (Description.Settings.AllowTearing && !Description.Settings.VSync)
            {
                NativeSwapChain.Present(0, PresentFlags.AllowTearing);
            }


            if (Description.Settings.VSync && !Description.Settings.AllowTearing)
            {
                NativeSwapChain.Present(1, PresentFlags.None);
            }


            BackBufferIndex = NativeSwapChain.CurrentBackBufferIndex;

        }     



        private void CreateSwapChainForDesktop()
        {
            SwapChainFlags Flags =  SwapChainFlags.None; // TODO: Defaul Full Screen ?


            if (Description.Settings.AllowTearing)
            {
                Flags = SwapChainFlags.AllowTearing;

                Description.Settings.VSync = false;
            }



            if (Description.Settings.VSync)
            {
                Flags = SwapChainFlags.None;

                Description.Settings.AllowTearing = false;
            }


            ModeDescription BackBufferDesc = new ModeDescription()
            {
                Width = Description.BackBufferWidth,
                Height = Description.BackBufferHeight,
                Format = Format.R8G8B8A8_UNorm,
                RefreshRate = new Rational()
                {
                    Numerator = 60,
                    Denominator = 1
                },
                ScanlineOrdering = ModeScanlineOrder.Unspecified,
                Scaling = ModeScaling.Unspecified,
            };

            SampleDescription sampleDescription = new SampleDescription()
            {
                Count = 1,
                Quality = 0
            };

            


            SwapChainDescription swapChainDesc = new SwapChainDescription()   
            {
                BufferCount = bufferCount,                                      // Set to a single back buffer.
                BufferDescription = BackBufferDesc,                             // Set the width and height of the back buffer.
                BufferUsage = Usage.RenderTargetOutput,                         // Set the usage of the back buffer.
                OutputWindow = Description.DeviceHandle,                        // Set the handle for the window to render to.
                SampleDescription = sampleDescription,                          // Turn multisampling off.
                Windowed = true,                                                // Set to full screen or windowed mode.
                Flags = Flags,                                                  // Don't set the advanced flags.
                SwapEffect = SwapEffect.FlipDiscard,                            // Discard the back buffer content after presenting.

                
            };
            





            IDXGIFactory4 Factory = GraphicsDevice.NativeAdapter.NativeFactory;

            IDXGISwapChain swapChain = Factory.CreateSwapChain(GraphicsDevice.NativeDirectCommandQueue.Queue, swapChainDesc);

            if (Description.Settings.Fullscreen)
            {
                // Before fullscreen switch
                swapChain.ResizeTarget(BackBufferDesc);

                // Switch to full screen
                swapChain.SetFullscreenState(true, default);

                // This is really important to call ResizeBuffers AFTER switching to IsFullScreen 
                swapChain.ResizeBuffers(bufferCount, Description.BackBufferWidth, Description.BackBufferHeight, Format.R8G8B8A8_UNorm, Flags);
            }



            NativeSwapChain = swapChain.QueryInterface<IDXGISwapChain3>();
        }
    }
}

File: Src/Xultaik.Graphics/SwapChain.cpp
#include "SwapChain.h"

#include "ConvertExtensions.h"
#include "GraphicsDevice.h"

#include <stdexcept>

namespace Xultaik::Graphics
{
    SwapChain::SwapChain(GraphicsDevice* graphicsDevice, HWND windowHandle, const RenderDescriptor& descriptor)
    {
        Microsoft::WRL::ComPtr<IDXGIFactory6> factory;
        if (FAILED(CreateDXGIFactory2(0, IID_PPV_ARGS(&factory))))
        {
            throw std::runtime_error("CreateDXGIFactory2 failed.");
        }

        DXGI_SWAP_CHAIN_DESC1 swapChainDesc{};
        swapChainDesc.Width = descriptor.width;
        swapChainDesc.Height = descriptor.height;
        swapChainDesc.Format = ConvertExtensions::ToDxgi(descriptor.format);
        swapChainDesc.SampleDesc.Count = 1;
        swapChainDesc.BufferUsage = DXGI_USAGE_RENDER_TARGET_OUTPUT;
        swapChainDesc.BufferCount = FrameCount;
        swapChainDesc.SwapEffect = DXGI_SWAP_EFFECT_FLIP_DISCARD;
        swapChainDesc.AlphaMode = DXGI_ALPHA_MODE_UNSPECIFIED;
        swapChainDesc.Scaling = DXGI_SCALING_STRETCH;
        swapChainDesc.Flags = descriptor.vsync ? 0u : DXGI_SWAP_CHAIN_FLAG_ALLOW_TEARING;

        Microsoft::WRL::ComPtr<IDXGISwapChain1> swapChain1;
        if (FAILED(factory->CreateSwapChainForHwnd(
                graphicsDevice->DirectQueue(),
                windowHandle,
                &swapChainDesc,
                nullptr,
                nullptr,
                &swapChain1)))
        {
            throw std::runtime_error("CreateSwapChainForHwnd failed.");
        }

        if (FAILED(swapChain1.As(&m_nativeSwapChain)))
        {
            throw std::runtime_error("SwapChain cast failed.");
        }

        factory->MakeWindowAssociation(windowHandle, DXGI_MWA_NO_ALT_ENTER);

        for (std::uint32_t index = 0; index < FrameCount; ++index)
        {
            if (FAILED(m_nativeSwapChain->GetBuffer(index, IID_PPV_ARGS(m_backBuffers[index].ReleaseAndGetAddressOf()))))
            {
                throw std::runtime_error("SwapChain GetBuffer failed.");
            }
        }
    }

    void SwapChain::Present(bool vsync)
    {
        m_nativeSwapChain->Present(vsync ? 1u : 0u, 0);
    }

    std::uint32_t SwapChain::CurrentBackBufferIndex() const noexcept
    {
        return m_nativeSwapChain->GetCurrentBackBufferIndex();
    }

    ID3D12Resource* SwapChain::BackBuffer(std::uint32_t index) const noexcept
    {
        return m_backBuffers[index].Get();
    }

    IDXGISwapChain4* SwapChain::Native() const noexcept
    {
        return m_nativeSwapChain.Get();
    }
}
